internal/elasticsearch/enrich/policy.go

```diff
@@ -0,0 +1,184 @@
+			Type:        schema.TypeString,
+			Required:    true,
+			ForceNew:    true,
+		},
+		"policy_type": {
+			Description:  "The type of enrich policy, can be one of geo_match, match, range.",
+			Type:         schema.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringInSlice([]string{"geo_match", "match", "range"}, false),
+		},
+		"indices": {
+			Description: "Array of one or more source indices used to create the enrich index.",
+			Type:        schema.TypeSet,
+			Required:    true,
+			ForceNew:    true,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+		"match_field": {
+			Description: "Field in source indices used to match incoming documents.",
+			Type:        schema.TypeString,
+			Required:    true,
+			ForceNew:    true,
+			ValidateFunc: validation.All(
+				validation.StringLenBetween(1, 255),
+			),
+		},
+		"enrich_fields": {
+			Description: "Fields to add to matching incoming documents. These fields must be present in the source indices.",
+			Type:        schema.TypeSet,
+			Required:    true,
+			ForceNew:    true,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+		"query": {
+			Description:  "Query used to filter documents in the enrich index. The policy only uses documents matching this query to enrich incoming documents. Defaults to a match_all query.",
+			Type:         schema.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsJSON,
+		},
+		"execute": {
+			Description: "Whether to call the execute API function in order to create the enrich index.",
+			Type:        schema.TypeBool,
+			Optional:    true,
+			ForceNew:    true,
+			Default:     true,
+		},
+	}
+
+	utils.AddConnectionSchema(policySchema)
+
+	return &schema.Resource{
+		Description: "Managing Elasticsearch enrich policies, see: https://www.elastic.co/guide/en/elasticsearch/reference/current/enrich-apis.html",
+
+		CreateContext: resourceEnrichPolicyPut,
+		UpdateContext: resourceEnrichPolicyPut,
+		ReadContext:   resourceEnrichPolicyRead,
+		DeleteContext: resourceEnrichPolicyDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: policySchema,
+	}
+}
+
+func resourceEnrichPolicyRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, diags := clients.NewApiClient(d, meta)
+	if diags.HasError() {
+		return diags
+	}
+
+	compName, diags := clients.CompositeIdFromStr(d.Id())
+	if diags.HasError() {
+		return diags
+	}
+	policy, diags := elasticsearch.GetEnrichPolicy(ctx, client, compName.ResourceId)
+	if policy == nil {
+		tflog.Warn(ctx, fmt.Sprintf(`Enrich policy "%s" not found, removing from state`, compName.ResourceId))
+		d.SetId("")
+		return diags
+	}
+	if diags == nil {
+		tflog.Warn(ctx, fmt.Sprintf(`Diags nil for resource ID "%s"`, compName.ResourceId))
+		return diags
+	}
```

Comment:
```suggestion
```

Why do we care about the distinction between nil/empty diags here?

Reply:
`elasticsearch.GetEnrichPolicy` can return nil when http.StatusNotFound which would make the next line `if diags.HasError()` throw some kind of null dereference error. I could make `elasticsearch.GetEnrichPolicy` return a `diag.FromErr(fmt.Error)` instead for that situation.

From looking at many other resources ex: [index.GetIndex()](https://github.com/elastic/terraform-provider-elasticstack/blob/main/internal/clients/elasticsearch/index.go#L317-L319) and it's use [resource](https://github.com/elastic/terraform-provider-elasticstack/blob/main/internal/elasticsearch/index/index.go#L916-L917) it combines the `diags != nil` test with the `model != nil` test. I could follow that pattern instead.

Reply:
> I could follow that pattern instead.

👍